### PR TITLE
Add docs to `patroni.dcs.__init__.py`

### PR DESCRIPTION
**Pull request by matthbakeredb**
Also, made some small code changes to satisfy formatting and pylint.

### CODE REVIEW

File: patroni/dcs/__init__.py [hunk 1]
```diff
@@ -1156,106 +1570,158 @@ def manual_failover(self, leader: Optional[str], candidate: Optional[str],
...
     @staticmethod
     def sync_state(leader: Optional[str], sync_standby: Optional[Collection[str]]) -> Dict[str, Any]:
-        """Build sync_state dict.
-        The sync_standby key being kept for backward compatibility.
-        :param leader: name of the leader node that manages /sync key
-        :param sync_standby: collection of currently known synchronous standby node names
-        :returns: dictionary that later could be serialized to JSON or saved directly to DCS
+        """Build ``sync_state`` dictionary.
+
+        The *sync_standby* key is kept for backward compatibility.
```
Comment (CyberDem0n) [inline patroni/dcs/__init__.py:1661]: I think we wanted to say that we don't use plural (`sync_standbys`) for backward compatibility.
```suggestion
```

File: patroni/dcs/__init__.py [hunk 2]
```diff
@@ -172,28 +193,46 @@ def get_dcs(config: Union['Config', Dict[str, Any]]) -> 'AbstractDCS':
 
 class Member(NamedTuple):
     """Immutable object (namedtuple) which represents single member of PostgreSQL cluster.
+
+    .. note::
+        There are two mandatory keys in a data:
+
+        ``conn_url``: connection string containing host, user and password which could be used to access this member.
+        ``api_url``: REST API url of patroni instance
```
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:201]: "Noteworthy"?

File: patroni/dcs/__init__.py [hunk 3]
```diff
@@ -172,28 +193,46 @@ def get_dcs(config: Union['Config', Dict[str, Any]]) -> 'AbstractDCS':
...
-        >>> Member.from_node(-1, '', '', '{"conn_url": "postgres://foo@bar/postgres"}') is not None
-        True
-        >>> Member.from_node(-1, '', '', '{')
-        Member(version=-1, name='', session='', data={})
+        """Factory method for instantiating :class:`Member` from a JSON serialised string or object.
+
+        :param version: modification version of a given member key in a Configuration Store.
+        :param name: name of PostgreSQL cluster member.
+        :param session: either session id or just ttl in seconds.
+        :param value: dictionary containing arbitrary data i.e. ``conn_url``, ``api_url``, ``xlog_location``, ``state``,
```
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:224]: Hmmn right, I remember looking at that code too. I'll change it.

File: patroni/dcs/__init__.py [hunk 4]
```diff
@@ -172,28 +193,46 @@ def get_dcs(config: Union['Config', Dict[str, Any]]) -> 'AbstractDCS':
...
+            ``role``, ``tags``, etc...
```
Comment (matthbakeredb) [inline patroni/dcs/__init__.py:225]:
```suggestion
        :param value: JSON encoded string containing arbitrary data i.e. ``conn_url``, ``api_url``, 
                      ``xlog_location``, ``state``, ``role``, ``tags``, etc. OR a connection URL
                      starting with ``postgres://``.
```

File: patroni/dcs/__init__.py [hunk 5]
```diff
@@ -302,84 +372,145 @@ class RemoteMember(Member):
...
         return self.member.conn_url
 
     @property
     def data(self) -> Dict[str, Any]:
+        """Data value of the :class:`Member` instance."""
         return self.member.data
 
     @property
     def timeline(self) -> Optional[int]:
+        """Timeline value of :attr:`~Leader.data`."""
```
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:437]: Yes that might be copypasta
```suggestion
        """Timeline value of :attr:`~Member.data`."""
```

File: patroni/dcs/__init__.py [hunk 6]
```diff
@@ -302,84 +372,145 @@ class RemoteMember(Member):
...
 
     @property
     def checkpoint_after_promote(self) -> Optional[bool]:
-        """
-        >>> Leader(1, '', Member.from_node(1, '', '', '{"version":"z"}')).checkpoint_after_promote
+        """Determine whether a checkpoint should be made for this leader after promotion.
+
+        :returns: ``True`` if the role is ``master`` or ``primary`` and ``checkpoint_after_promote`` is not set,
+                 ``False`` if not a ``master`` or ``primary`` or if the checkpoint has already been made.
+                 If the version of Patroni is older than 1.5.6, return ``None``.
```
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:446]: So it's a truth check? checkpoint **has** occurred?
```suggestion
        """Determine whether a checkpoint has been made for this leader after promotion.

        :returns: ``True`` if the role is ``master`` or ``primary`` and ``checkpoint_after_promote`` is not set,
                 ``False`` if not a ``master`` or ``primary`` or if the checkpoint hasn't been made.
                 If the version of Patroni is older than 1.5.6, return ``None``.
```

File: patroni/dcs/__init__.py [hunk 7]
```diff
@@ -585,50 +795,87 @@ class Cluster(NamedTuple):
...
+        """Produce an empty :class:`Cluster` instance."""
         return Cluster(None, None, None, 0, [], None, SyncState.empty(), None, None, None)
 
     def is_empty(self):
-        return self.initialize is None and self.config is None and self.leader is None and self.last_lsn == 0\
-            and self.members == [] and self.failover is None and self.sync.version is None\
-            and self.history is None and self.slots is None and self.failsafe is None and self.workers == {}
+        """Validate definition of all attributes of this :class:`Cluster` instance.
+
+        :returns: ``True`` if the current :class:`Cluster` has unpopulated attributes.
```
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:804]: See my suggestion https://github.com/zalando/patroni/pull/2777/files#r1276421207

File: patroni/dcs/__init__.py [hunk 8]
```diff
@@ -585,50 +795,87 @@ class Cluster(NamedTuple):
...
+        """
+        return all((self.initialize is None, self.config is None, self.leader is None, self.last_lsn == 0,
+                    self.members == [], self.failover is None, self.sync.version is None,
+                    self.history is None, self.slots is None, self.failsafe is None, not self.workers))
 
     def __len__(self) -> int:
+        """Implement ``len`` function capability."""
```
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:811]: Ah that makes sense, so this is more of a "is defined" test? `len` doesn't sound right for that, in which case yes I agree.

Would you want to make that a separate changeset/PR?

File: patroni/dcs/__init__.py [hunk 9]
```diff
@@ -800,25 +1062,48 @@ def should_enforce_hot_standby_feedback(self, my_name: str, nofailover: bool, ma
...
-        0
-        >>> Cluster(0, 0, 0, 0, 0, 0, 0, TimelineHistory.from_node(1, '[]'), 0, None).timeline
-        1
-        >>> Cluster(0, 0, 0, 0, 0, 0, 0, TimelineHistory.from_node(1, '[["a"]]'), 0, None).timeline
-        0
+        """Get the cluster history index from the :attr:`~Cluster.history`.
+
+        :returns: If the recorded history is empty assume timeline is ``1``, if it is not defined or the stored history
+                  is not formatted as expected ``0`` is returned and an error will be logged.
+                  Otherwise, the first number stored (indexed starting at 1) is returned.
```
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:1088]: Ok, thanks for explaining.

File: patroni/dcs/__init__.py [hunk 10]
```diff
@@ -850,6 +1143,7 @@ def wrapper(*args: Any, **kwargs: Any):
 
 
 class AbstractDCS(abc.ABC):
+    """Abstract representation of DCS for preparing and loading :class:`Cluster` objects."""
```
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:1146]: OK, I'll have a look at this and flesh it out a bit.

File: patroni/dcs/__init__.py [hunk 11]
```diff
@@ -1006,15 +1341,28 @@ def __get_patroni_cluster(self, path: Optional[str] = None) -> Cluster:
         return cluster
 
     def is_citus_coordinator(self) -> bool:
+        """:class:`Cluster` instance has a Citus Coordinator group ID.
+
+        :returns: ``True`` if this object's Citus group ID matches the default Citus Coordinator group ID.
```
Comment (matthbakeredb) [inline patroni/dcs/__init__.py:1346]:
```suggestion
        :returns: ``True`` if the given node is running as Citus Coordinator (``group=0``).
```

File: patroni/dcs/__init__.py [hunk 12]
```diff
@@ -1006,15 +1341,28 @@ def __get_patroni_cluster(self, path: Optional[str] = None) -> Cluster:
...
+        """
         return self._citus_group == str(CITUS_COORDINATOR_GROUP_ID)
 
     def get_citus_coordinator(self) -> Optional[Cluster]:
+        """Load the cluster for the Citus Coordinator node.
+
```
Comment (matthbakeredb) [inline patroni/dcs/__init__.py:1352]:
```suggestion

          .. note::
              This method is only executed on the worker nodes (``group!=0``) to find the coordinator.

```

File: patroni/dcs/__init__.py [hunk 13]
```diff
@@ -1006,15 +1341,28 @@ def __get_patroni_cluster(self, path: Optional[str] = None) -> Cluster:
         return cluster
 
     def is_citus_coordinator(self) -> bool:
+        """:class:`Cluster` instance has a Citus Coordinator group ID.
+
+        :returns: ``True`` if this object's Citus group ID matches the default Citus Coordinator group ID.
+        """
```
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:1347]: See https://github.com/zalando/patroni/pull/2777/files#r1276434611

File: patroni/dcs/__init__.py [hunk 14]
```diff
@@ -1024,6 +1372,17 @@ def _get_citus_cluster(self) -> Cluster:
         return cluster
 
     def get_cluster(self, force: bool = False) -> Cluster:
+        """Retrieve an appropriate cached or fresh view of DCS.
```
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:1375]: Ah ok, I added an extra note comment https://github.com/zalando/patroni/pull/2777/files#r1276442461

File: patroni/dcs/__init__.py [hunk 15]
```diff
@@ -1043,32 +1402,51 @@ def get_cluster(self, force: bool = False) -> Cluster:
...
 
-        :param last_lsn: absolute WAL LSN in bytes
-        :returns: `!True` on success."""
+        :param last_lsn: absolute WAL LSN in bytes.
+
+        :returns: ``True`` if successfully committed to DCS.
+        """
 
     def write_leader_optime(self, last_lsn: int) -> None:
+        """Write value for WAL LSN to ``optime/leader`` key in DCS.
```
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:1425]: So perhaps a note to say this is only used on older Patroni versions?

(When I saw that before I did wonder why you need to keep that around. Is it just for clusters upgrading from old versions?)

File: patroni/dcs/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:447]: Yes, this method returns `True` when checkpoint after promote has occurred on the given primary node.

File: patroni/dcs/__init__.py [hunk 16]
```diff
@@ -402,21 +533,37 @@ def from_node(version: _Version, value: Union[str, Dict[str, str]]) -> 'Failover
         return Failover(version, data.get('leader'), data.get('member'), data.get('scheduled_at'))
 
     def __len__(self) -> int:
+        """Implement ``len`` function capability."""
```
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:537]: I think it would be worth to mention that this method is called when something like `bool(cluster.failover)` or `if cluster.failover` is executed.

The reason it was done this way because Patroni need to create/update/delete failover key atomically, but ZooKeeper implements atomic operations it very weirdly. In ZooKeeper these operations are based on version of a given ZNode, which is reset to 0 every time when ZNode is dropper and created. That is, we have a /failover key with version=0 and want to remove it, but meanwhile some other thread removed and added it back. As a result we will be removing not a key we wanted to remove.

To mitigate this behavior Patroni just empties the /failover key instead of removing it. When we read the cluster state from DCS, the Failover object will be created, because we need to know its version for the next update.

Without `__len__` we would have to write something like `if cluster.failover and (cluster.failover.leader or cluster.failover.candidate)` most of the time, and now we can just do `if cluster.failover`, which will be evaluated to `False` if object is "empty".

File: patroni/dcs/__init__.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:1379]: No. ZooKeeper implements caching inside zookeeper.py.

`self._cluster = cluster` -- this cache exists mainly for REST API, in order to not put additional stress on DCS while processing health-check requests.

File: patroni/dcs/__init__.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:1430]: Actually, this method is still called from ha.py on demote or on Patroni shutdown (which also demotes postgres).
It was kept in order to avoid calling write_status() directly, because write_status() requires dict object to be passed and I wanted to abstract it away.

But In any case it first calls write_status() method and only the last one decides whether it should also update /optime/leader

File: patroni/dcs/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:201]: When Patroni updates the member key it always writes conn_url and api_url, but what if the key was updated by someone else? We want to be resilient for such unexpected scenarios.

[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:201]: I don't understand this. So `Member` is read from DCS during `dcs.touch_member`, and if it's different to `data` passed into it, then the data constructed in `ha.touch_member` is written (looking at the etcd3 implementation). The values for each of these 2 parameters are taken from `state_handler.connection_string` and `api.connection_string` in the `ha.touch_member`. I don't see where the existing data held in DCS is kept/merged.

Do you mean someone who wrote to the DCS `/member/member` key outside of Patroni? Who would do that? And why would that be considered something to honour?

I am thinking that for this comment it's worth mentioning that those keys are assembled from active connection information when data is written to DCS.

Also, why doesn't touch_member create a `Member` instance instead of constructing a `Dict`?

Thanks for any pointers!

File: patroni/dcs/__init__.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:537]: Thanks that makes sense!

File: patroni/dcs/__init__.py [hunk 17]
```diff
@@ -402,21 +534,37 @@ def from_node(version: _Version, value: Union[str, Dict[str, str]]) -> 'Failover
...
```
Comment (matthbakeredb) [inline patroni/dcs/__init__.py:537]:
```suggestion
        """Implement ``len`` function capability.
        
        .. note::
           This magic method aids in the evaluation of "emptiness" of a ``Failover`` instance. For example:
           
           >>> failover = Failover.from_node(1, None)
           >>> len(failover)
           0
           >>> assert bool(failover) is False
           
           >>> failover = Failover.from_node(1, {"leader": "cluster_leader"})
           >>> len(failover)
           1
           >>> assert bool(failover) is True
        
           This makes it easier to write ``if cluster.failover`` rather than the longer statement.
        
        """
```

File: patroni/dcs/__init__.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:537]: See suggestion https://github.com/zalando/patroni/pull/2777/files#r1279176340

File: patroni/dcs/__init__.py [hunk 18]
```diff
@@ -585,50 +796,87 @@ class Cluster(NamedTuple):
...
+        """Validate definition of all attributes of this :class:`Cluster` instance.
+
+        :returns: ``True`` if the current :class:`Cluster` has unpopulated attributes.
+        """
+        return all((self.initialize is None, self.config is None, self.leader is None, self.last_lsn == 0,
+                    self.members == [], self.failover is None, self.sync.version is None,
+                    self.history is None, self.slots is None, self.failsafe is None, not self.workers))
 
     def __len__(self) -> int:
+        """Implement ``len`` function capability."""
```
Comment (matthbakeredb) [inline patroni/dcs/__init__.py:812]:
```suggestion
        """Implement ``len`` function capability.

        .. note::
           This magic method aids in the evaluation of "emptiness" of a ``Cluster`` instance. For example:

           >>> cluster = Cluster.empty()
           >>> len(cluster)
           0
           >>> assert bool(cluster) is False

           >>> cluster = Cluster(None, None, None, 0, [1, 2, 3], None, SyncState.empty(), None, None, None)
           >>> len(cluster)
           1
           >>> assert bool(cluster) is True

           This makes it easier to write ``if cluster`` rather than the longer statement.

        """
```

File: patroni/dcs/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:812]: As per the other suggestion, see this one for `Cluster` https://github.com/zalando/patroni/pull/2777/files#r1279186080

File: patroni/dcs/__init__.py (hunk 17 above)
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:537]: Looks good

File: patroni/dcs/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:201]: > So Member is read from DCS during dcs.touch_member

No, touch_member() write member key. Reading happens somewhere inside in get_cluster().

> and if it's different to data passed into it, then the data constructed in ha.touch_member is written (looking at the etcd3 implementation)

This is unrelated and just implementation detail of very specific DCS, because the member key is connected with the session and will not expire while the session exists. Therefore we can skip the update if value didn't changed.
But, with for example etcd v2 the member key is updated every time.

> Do you mean someone who wrote to the DCS /member/member key outside of Patroni? Who would do that? And why would that be considered something to honour?

Exactly, something different from Patroni could change arbitrary keys in DCS and we better be resilient to it.

> Also, why doesn't touch_member create a Member instance instead of constructing a Dict?

Because touch_member() only updates one key that represents current node in DCS. When we read the cluster state we are reading **all** member keys, which happens on all nodes.

File: patroni/dcs/__init__.py [hunk 19]
```diff
@@ -1024,6 +1376,18 @@ def _get_citus_cluster(self) -> Cluster:
         return cluster
 
     def get_cluster(self, force: bool = False) -> Cluster:
+        """Retrieve an appropriate cached or fresh view of DCS.
+
+        .. note::
+            Stores copy of time, status and failsafe values for comparison in DCS update decisions.
+            This works around the unique operational requirement of Zookeeper to make extra queries for keys.
```
Comment (matthbakeredb) [inline patroni/dcs/__init__.py:1383]:
```suggestion
            Caching is mostly required to avoid overhead placed on the REST API, due to the nature of how
            Zookeeper makes multiple requests are made in comparison to other DCS implementations.
```

File: patroni/dcs/__init__.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:1379]: Um OK, so that suggestion is incorrect? Confused... how about this? https://github.com/zalando/patroni/pull/2777/files#r1279198612

File: patroni/dcs/__init__.py [hunk 20]
```diff
@@ -1043,32 +1407,51 @@ def get_cluster(self, force: bool = False) -> Cluster:
...
+        :param last_lsn: absolute WAL LSN in bytes.
+
+        :returns: ``True`` if successfully committed to DCS.
+        """
 
     def write_leader_optime(self, last_lsn: int) -> None:
+        """Write value for WAL LSN to ``optime/leader`` key in DCS.
+
+        :param last_lsn: absolute WAL LSN in bytes.
+        """
```
Comment (matthbakeredb) [inline patroni/dcs/__init__.py:1433]:
```suggestion
        """Write value for WAL LSN to ``optime/leader`` key in DCS.

        .. note::
            This method abstracts away the required data structure of :meth:`~Cluster.write_status`, so it
            is not needed in the caller. However, the ``optime`` are only written in :meth:`~Cluster.write_status`
            when the member Patroni version is old enough to require it.

        :param last_lsn: absolute WAL LSN in bytes.
        """
```

File: patroni/dcs/__init__.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:1430]: Does this addition make sense? https://github.com/zalando/patroni/pull/2777/files#r1279211995

File: patroni/dcs/__init__.py (hunk 19 above)
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:1383]: I wouldn't mention Zookeeper or any other DCS at all here.

TL;DR - there are two levels of caches:
1. `self._cluster` which is accessed via `cluster` property from REST API and cached only for `ttl` seconds.
2. sometimes DCS are have their own caches:
  - zookeeper does it for the entire Cluster
  - K8s and Etcd3 have caches for all keys associated with the cluster

The second type of caches exists to avoid reading something from DCS on every heart beat (when it is possible), and we shouldn't reveal here any implementation details.

File: patroni/dcs/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:201]: > How are we resilient to it? 

Until Patroni on respective node overwrites it.
That is, we don't want to suddenly crash because can't deserialize something.
But, we still create the Member object because the node that performs update of the member key sometimes should know that the key exists and should be removed before creating a new one.
For example in Zookeeper and Consul keys could be bound to a session and it is not possible (or there is no reliable way of doing) to change the session for the key.

File: patroni/dcs/__init__.py [hunk 21]
```diff
@@ -585,50 +813,106 @@ class Cluster(NamedTuple):
 
     @staticmethod
     def empty() -> 'Cluster':
-        return Cluster(None, None, None, 0, [], None, SyncState.empty(), None, None, None)
+        """Produce an empty :class:`Cluster` instance."""
+        return Cluster(None, None, None, 0, [], None, SyncState.empty(), None, None, None, {})
```
Comment (matthbakeredb) [inline patroni/dcs/__init__.py:817]: @CyberDem0n looks like there was a bug here? It was preventing the doctest from passing because some Citus tests write to cluster.workers. Should I create a separate PR for this?

Comment (matthbakeredb) [reply patroni/dcs/__init__.py:817]: Bug is there are in fact 11 args to `Cluster` not 10...

Comment (matthbakeredb) [reply patroni/dcs/__init__.py:817]: Hmmn seems to have broken tests, which passed locally. :thinking: 

https://github.com/zalando/patroni/actions/runs/5717044205/job/15489954467

Comment (CyberDem0n) [reply patroni/dcs/__init__.py:817]: Yeah, with workers it is an interesting corner case.
It has a default value `{}`, which is mutated :(
The problem is introduced in #2652, when I changed `namedtuple` with overridden `__new__` method to `NamedTuple`.
I'll open a separate PR with the fix.

File: patroni/dcs/__init__.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:1466]: Thank you!

File: patroni/dcs/__init__.py (hunk 19 above)
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:1419]: Updated it to a simpler form. Thanks for the explanation.

File: patroni/dcs/__init__.py (hunk 21 above)
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:817]: Ah, thanks for taking that. I'll rebase when it's merged.

File: patroni/dcs/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:201]: Ah right, got it. So you're referring to the read of DCS being resilient to changes UNTIL we overwrite it with sane values.

[the remark this answers is not in the document]
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:201]: Anything is possible. Maybe the broken member key will never be updated, we just need to make sure that it will not affect normal cluster operations. For example we don't want to create a replication slot for it.

File: patroni/dcs/__init__.py (hunk 19 above)
Comment (CyberDem0n) [reply patroni/dcs/__init__.py:1419]: :+1:

File: patroni/dcs/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:201]: How about this:

```suggestion
    .. note::
        These two keys in data are always written to the DCS, but care is taken to maintain consistency and resilience
        from data that is read:
        
        ``conn_url``: connection string containing host, user and password which could be used to access this member.
        ``api_url``: REST API url of patroni instance
```

File: patroni/dcs/__init__.py [hunk 22]
```diff
@@ -898,6 +1281,12 @@ def __init__(self, config: Dict[str, Any]) -> None:
         self.event = Event()
 
     def client_path(self, path: str) -> str:
```
Comment (CyberDem0n) [inline patroni/dcs/__init__.py:1283]: Yeah, now the *path* word is very confusing in this context. Originally we used `path` only because of actual hierarchical structures in Etcd v2 and Zookeeper.

It is all about building key names in DCS from the following information:
- namespace (key prefix)
- scope (cluster name)
- key name within the scope

At the end we get something like the following.
1. for etcd, etcd3, and raft:
```
/{namespace}/{scope}/config
/{namespace}/{scope}/leader
/{namespace}/{scope}/members/{my_name}
...
```

2. for consul and zookeeper:
```
{namespace}/{scope}/config
{namespace}/{scope}/leader
{namespace}/{scope}/members/{my_name}
...
```

3. something totally different for K8s (all K8s objects should be created in the same namespace that matches `kuberneted.namespace` parameter):
```
{scope}-leader  # leader configmap
{scope}  # leader endpoint

{scope}-{config}  # config key configmap or endpoint
...
```
Pods (that map to members) may have arbitrary names and identified by their labels.


Now the word `key` would be more appropriate. But we can do it separately.

File: patroni/dcs/__init__.py [hunk 23]
```diff
@@ -864,6 +1190,62 @@ def wrapper(*args: Any, **kwargs: Any):
...
+    Functional methods that perform construction of complex data objects:
+
+        * :meth:`~AbstractDCS._cluster_loader`:
+            method which processes the structure of data stored in the DCS used to build the :class:`Cluster` object
+            with all relevant associated data.
+        * :meth:`~AbstractDCS._citus_cluster_loader`:
+            Similar to above but specifically representing Citus group and workers information.
+        * :meth:`~AbstractDCS._load_cluster`:
+            main method for calling specific ``loader`` method to build the :class:`Cluster` object representing the
+            state and topology of the cluster.
```
Comment (CyberDem0n) [inline patroni/dcs/__init__.py:1207]: They all time critical. If didn't succeed during `retry_timeout` the primary is demoted because of lack of access to DCS.

Comment (matthbakeredb) [reply patroni/dcs/__init__.py:1207]: Thanks for these comments on this section. I'll bundle them together in one shuffle commit.

File: patroni/dcs/__init__.py [hunk 24]
```diff
@@ -302,84 +374,145 @@ class RemoteMember(Member):
 
     @classmethod
     def from_name_and_data(cls, name: str, data: Dict[str, Any]) -> 'RemoteMember':
+        """Factory method to construct instance from given *name* and *data*.
+
+        :param name: name of the remote member.
+        :param data: dictionary of member information.
```
Comment (barthisrael) [inline patroni/dcs/__init__.py:380]: What do we expecte `data` to contain?

File: patroni/dcs/__init__.py [hunk 25]
```diff
@@ -402,21 +535,54 @@ def from_node(version: _Version, value: Union[str, Dict[str, str]]) -> 'Failover
         return Failover(version, data.get('leader'), data.get('member'), data.get('scheduled_at'))
 
     def __len__(self) -> int:
+        """Implement ``len`` function capability.
+
+        .. note::
+           This magic method aids in the evaluation of "emptiness" of a ``Failover`` instance. For example:
```
Comment (barthisrael) [inline patroni/dcs/__init__.py:541]:
```suggestion
           This magic method aids in the evaluation of "emptiness" of a :class:`Failover` instance. For example:
```

File: patroni/dcs/__init__.py [hunk 26]
```diff
@@ -527,25 +732,45 @@ def matches(self, name: Optional[str], check_leader: bool = False) -> bool:
...
     @staticmethod
     def from_node(version: _Version, value: str) -> 'TimelineHistory':
-        """
-        >>> h = TimelineHistory.from_node(1, 2)
-        >>> h.lines
-        []
+        """Parse the given JSON serialized string as a list of timeline history lines.
+
+        :param version: version number
+        :param value: JSON serialized string.
```
Comment (barthisrael) [inline patroni/dcs/__init__.py:762]: Any expected content?

File: patroni/dcs/__init__.py [hunk 27]
```diff
@@ -906,112 +1295,147 @@ def client_path(self, path: str) -> str:
...
     @property
     @abc.abstractmethod
     def ttl(self) -> int:
-        """Get new ttl value"""
+        """Get current ``ttl`` value."""
 
     @abc.abstractmethod
     def set_retry_timeout(self, retry_timeout: int) -> None:
-        """Set the new value for retry_timeout"""
+        """Set the new value for retry_timeout."""
```
Comment (barthisrael) [inline patroni/dcs/__init__.py:1362]:
```suggestion
        """Set the new value for *retry_timeout*."""
```

File: patroni/dcs/__init__.py [hunk 28]
```diff
@@ -1059,32 +1512,57 @@ def get_cluster(self, force: bool = False) -> Cluster:
...
+            is old enough to require it (i.e. the old Patroni version doesn't understand the new format).
+
+        :param last_lsn: absolute WAL LSN in bytes.
+        """
         self.write_status({self._OPTIME: last_lsn})
 
     @abc.abstractmethod
     def _write_status(self, value: str) -> bool:
-        """write current WAL LSN and confirmed_flush_lsn of permanent slots into the `/status` key in DCS
+        """Write current WAL LSN and confirmed_flush_lsn of permanent slots into the ``/status`` key in DCS.
```
Comment (barthisrael) [inline patroni/dcs/__init__.py:1549]:
```suggestion
        """Write current WAL LSN and ``confirmed_flush_lsn`` of permanent slots into the ``/status`` key in DCS.
```

File: patroni/dcs/__init__.py [hunk 29]
```diff
@@ -1097,38 +1575,55 @@ def write_status(self, value: Dict[str, Any]) -> None:
...
+        :param value: failsafe topology serialized in JSON format.
+
+        :returns: ``True`` if successfully committed to DCS.
+        """
 
     def write_failsafe(self, value: Dict[str, str]) -> None:
-        if not (isinstance(self._last_failsafe, dict) and deep_compare(self._last_failsafe, value))\
+        """Write the ``/failsafe`` key in DCS.
+
+        :param value: dictionary value to set.
```
Comment (barthisrael) [inline patroni/dcs/__init__.py:1586]: Any expected content?

Comment (matthbakeredb) [reply patroni/dcs/__init__.py]: Updated

File: patroni/dcs/__init__.py (hunk 24 above)
Comment (matthbakeredb) [reply patroni/dcs/__init__.py]: updated

File: patroni/dcs/__init__.py (hunk 26 above)
Comment (matthbakeredb) [reply patroni/dcs/__init__.py]: Updated

**Comment by CyberDem0n** [comment]
After release there are now conflicts.

**Comment by matthbakeredb** [comment]
> After release there are now conflicts.

Rebased

**Comment by hughcapet** [comment]
Decreased coverage here as well:(
https://coveralls.io/builds/61766756/source?filename=patroni%2Fdcs%2F__init__.py#L399

File: tests/test_postgresql.py [hunk 30]
```diff
@@ -347,6 +347,7 @@ def test_write_postgresql_and_sanitize_auto_conf(self):
     def test_follow(self):
         self.p.call_nowait(CallbackAction.ON_START)
         m = RemoteMember('1', {'restore_command': '2', 'primary_slot_name': 'foo', 'conn_kwargs': {'host': 'bar'}})
+        self.assertEqual(m.disallowed_attr, None)
```
Comment (CyberDem0n) [inline tests/test_postgresql.py:350]: I wonder if declaring `RemoteMember.__getattr__()` as returning Optional[Any] would work better?

Comment (matthbakeredb) [reply tests/test_postgresql.py:350]: Do you mean in the test or in the code?

The extra `return None` is not part of typing tests. It comes from https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/inconsistent-return-statements.html

Comment (CyberDem0n) [reply tests/test_postgresql.py:350]: I mean that `return None` is redundant, it is returned implicitly.

Comment (matthbakeredb) [reply tests/test_postgresql.py:350]: Or

```
    def __getattr__(self, name: str) -> Any:
        """Dictionary style key lookup.

        :param name: key to lookup.

        :returns: value of *name* key in :attr:`~RemoteMember.data` if key *name* is in
                  :cvar:`~RemoteMember.ALLOWED_KEYS`, else ``None``.
        """
        return self.data.get(name) if name in RemoteMember.ALLOWED_KEYS else None
```

Comment (matthbakeredb) [reply tests/test_postgresql.py]: Removed both `return None` and test.

**Comment by hughcapet** [comment]
👍

**Comment by CyberDem0n** [comment]
:+1:
